File: USACO/2016-2017/Feb 2017/Plat/mincross.java
// Why Did the Cow Cross the Road

import java.util.*;
import java.io.*;

public class mincross {
	
	static int[] BIT;
	public static void main(String[] args) throws IOException {
		String file = "";
		// BufferedReader br = new BufferedReader(new FileReader(new File(file + ".in")));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		BIT = new int[N+1];
		int[] start = new int[N+1];
		int[][] loc = new int[N+1][2];
		
		for(int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			start[Integer.parseInt(st.nextToken())] = i;
		}
		
		loc[0][0] = -1;
		for(int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			int from = start[Integer.parseInt(st.nextToken())];
			loc[from][0] = i;
			loc[from][1] = from;
		}
		br.close();
		
		int[][] sortedLoc = Arrays.copyOf(loc, N+1);
		Arrays.sort(sortedLoc, (a, b) -> a[0] - b[0]);
		
		long intersections = 0;
		long minI = intersections;
		
		// build BIT
		for(int i = N; i >= 1; i--) {
			int from = sortedLoc[i][1];
			intersections += get(from-1);
			update(from, 1);
		}
		
		// shift second row
		for(int i = 1; i <= N; i++) {
			intersections -= sortedLoc[i][1] - 1;
			intersections += N - sortedLoc[i][1];
			
			minI = Math.min(minI, intersections);
		}
		
		// shift first row
		for(int i = 1; i <= N; i++) {
			intersections -= loc[i][0] - 1;
			intersections += N - loc[i][0];
			
			minI = Math.min(minI, intersections);
		}
		
		//PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(new File(file + ".out"))));
		System.out.println(minI);
		//out.println();
		//out.close();
	}
	
	public static void update(int idx, int val) {
		while(idx < BIT.length) {
			BIT[idx] += val;
			idx += idx & -idx;
		}
	}
	
	public static int get(int idx) {
		int sum = 0;
		
		while(idx > 0) {
			sum += BIT[idx];
			idx -= idx & -idx;
		}
		
		return sum;
	}
}

File: USACO/2016-2017/Feb 2017/Plat/mincross.cpp
// Why Did the Cow Cross the Road

#include <iostream>
#include <algorithm>
#include <vector>

#define MAXN 100005
#define ll long long

using namespace std;

int BIT[MAXN];

void upd(int idx, int add) {
	while(idx < MAXN) {
		BIT[idx] += add;
		idx += idx & -idx;
	}
}

int get(int idx) {
	int sum = 0;
	while(idx > 0) {
		sum += BIT[idx];
		idx -= idx & -idx;
	}
	
	return sum;
}

bool comp(vector<int> a, vector<int> b) { return a[0] < b[0]; }

int main() {
	int N; cin >> N;
	
	int start[N+1];
	vector<vector<int>> loc(N+1);
	
	for(int i = 1; i <= N; i++) { 
		int a;
		cin >> a;
		start[a] = i;
	}
	
	loc[0].push_back(-1);
	for(int i = 1; i <= N; i++) {
		int a;
		cin >> a;
		loc[start[a]].push_back(i);
		loc[start[a]].push_back(start[a]);
	}
	
	vector<vector<int>> sortedLoc = loc;
	sort(sortedLoc.begin(), sortedLoc.end(), comp);
	
	// update BIT
	ll intersections = 0;
	
	for(int i = N; i >= 1; i--) {
		int from = sortedLoc[i][1];
		intersections += get(from-1);
		upd(from, 1);
	}
	
	ll minI = intersections;
	// shift 2nd row
	for(int i = 1; i <= N; i++) {
		intersections -= sortedLoc[i][1] - 1;
		intersections += N - sortedLoc[i][1];
		
		minI = min(minI, intersections);
	}
	
	// shift 1st row
	for(int i = 1; i <= N; i++) {
		intersections -= loc[i][0] - 1;
		intersections += N - loc[i][0];
		
		minI = min(minI, intersections);
	}
	
	cout << minI << endl;
}

